File: ch12_the_sparse_extended_information_filter/src/julia/motion.jl
function motion(v,α,dt,m₀,μ,ξ,Ω,graph)

    n = length(ξ)
    rnoise = diagm(0 => [8e-4,8e-4,1e-6])
    jcb = jacobian_motion(μ[3],v,α,dt)

    ψ = inv(jcb)-I
    ind = [[2*i+2;2*i+3] for i ∈ sort(m₀)]
    ind = vcat(ind...)
    ind = [1:3;ind]
    Ω₁ = Ω[ind,1:3]
    ψ₁ = Ω₁*ψ

    λ = zeros(n,n)
    λ[ind,1:3] = ψ₁
    λ[1:3,ind] .+= ψ₁'
    λ[1:3,1:3] .+= ψ'*Ω[1:3,1:3]*ψ
    Φ = Ω + λ

    Φₓ = Φ[1:3,ind]
    κ = zeros(n,n)
    κ[ind,ind] = Φₓ'*inv(inv(rnoise)+Φ[1:3,1:3])*Φₓ
    Ω .+= λ - κ

    δ = equation_motion(μ[3],v,α,dt)
    ξ[ind] .+= Ω[ind,1:3]*δ + (λ-κ)[ind,:]*μ
    μ[1:3] += δ

    if length(m₀)>0
        edges = [[i+1 j+1] for i ∈ m₀,j ∈ m₀ if i ≠ j]
        edges = vcat(edges...)
        if size(edges,1)>0
            sg = size(graph,1)
            adjacency = sparse(edges[:,1],edges[:,2],true,sg,sg)
            graph .|= adjacency
        end
    end
end

File: ch12_the_sparse_extended_information_filter/src/cpp/motion.cc
#include "seif.h"

void motion(float v, float α, float dt, std::vector<uint> &m0, vec &μ, vec &ξ, mat &Ω, SpMat<ushort> &Λ){
  size_t n = ξ.n_elem;
  mat rnoise = diagmat(vec({8e-4,8e-4,1e-6}));

  mat jcb = jacobian_motion(μ(2),v,α,dt);

  mat ψ = inv(jcb) - eye<mat>(3,3);
  uvec indx = {0,1,2};
  std::vector<uint> m1(m0);
  sort(m1.begin(),m1.end());

  uvec ind(3+2*m1.size());
  ind(uvec({0,1,2})) = uvec({0,1,2});
  uint j(0);
  for (auto iter=m1.begin(); iter != m1.end(); ++iter, j+=2) {
    ind(3+j) = 2**iter+1;
    ind(4+j) = 2**iter+2;
  }
  mat Ω₁ = Ω(ind,uvec(indx));
  mat ψ₁ = Ω₁*ψ;

  mat λ = zeros<mat>(n,n);
  λ(ind,indx) = ψ₁;
  λ(indx,ind) += ψ₁.t();
  λ(indx,indx) += ψ.t()*Ω(indx,indx)*ψ;
  mat Φ = Ω + λ;

  mat Φₓ = Φ(indx,ind);
  mat κ = zeros<mat>(n,n);
  κ(ind,ind) = Φₓ.t()*inv(inv(rnoise)+Φ(indx,indx))*Φₓ;
  Ω += λ - κ;

  vec δ = equation_motion(μ(2),v,α,dt);

  mat lk = λ-κ;
  ξ(ind) += Ω(ind,indx)*δ + lk.rows(ind)*μ;
  μ(span(0,2)) += δ;

  if (!m0.empty()) {
    umat edges = zeros<umat>(2,m1.size()*m1.size());
    uint i(0);
    for (auto iter=m1.begin(); iter != m1.end(); ++iter) {
      for (auto iter1 = m1.begin(); iter1 != m1.end(); ++iter1) {
        edges(0,i) = *iter;
        edges(1,i++) = *iter1;
      }
    }
    if (!edges.empty()) {
      SpMat<ushort> adjacency(edges,ones<Col<ushort>>(edges.n_cols),Λ.n_rows,Λ.n_cols);
      Λ += adjacency;
    }
  }
}
